core/recovery_test.go

```diff
@@ -86,53 +120,65 @@ func TestRecovery(t *testing.T) {
 	// states should be identical.
 	timestamp := time.Now()
 	var databaseDumps []string
-	for n := 1; ; n++ {
+	for n := int64(1); ; n++ {
+		ctx = context.Background()
+		t.Logf("Beginning run; will crash on %d-th SQL query", n)
 		// Create a new, fresh database using the schema file we created
 		// above. The database will be identical.
 		cloneURL, err := pgtest.CloneDB(ctx, dbURL)
 		if err != nil {
 			t.Fatal(err)
 		}
 
+		runCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
 		// Open a new handle to the same database but using a driver that
 		// will call our anonymous func to simulate crashes.
-		ch := make(chan error)
-		calls := 0
+		crashed := make(chan struct{})
+		completed := make(chan struct{})
+		var calls int64
 		wrappedDB := pgtest.WrapDB(t, cloneURL, func(q string) {
-			calls++
-
-			if calls == n {
-				t.Logf("crashing on query %d: %s\n", calls, q)
-				close(ch)        // let main goroutine know
+			num := atomic.AddInt64(&calls, 1)
+			if num == n {
+				t.Logf("crashing on query %d: %s\n", num, q)
+				cancel()         // cancel the context so other goroutines clean up too
+				close(crashed)   // let main goroutine know
 				runtime.Goexit() // crash this goroutine
 			}
 		})
 
-		ctx := context.Background()
 		go func() {
-			err := generateBlock(ctx, t, wrappedDB, timestamp, poolTxs)
-			ch <- err
+			err = generateBlock(runCtx, t, wrappedDB, timestamp, poolTxs)
+			if err != nil {
+				t.Errorf("generateBlock returned err %s", err)
+			}
+			cancel()
+			close(completed)
 		}()
 
 		// Wait for the goroutine to finish or get killed on the N-th query.
-		err, ok := <-ch
-		if err != nil {
-			t.Fatal(err)
+		select {
+		case <-crashed:
+			t.Log("goroutine crashed on query")
+		case <-completed:
+			t.Log("goroutine completed successfully")
 		}
-		if ok {
+
+		if atomic.LoadInt64(&calls) < n {
 			// The driver never crashed the goroutine, so n is now greater than
```

Comment:
This old comment should be updated the reflect the new code: "`calls` never reached `n` so the goroutine completed without simulating a crash"